src/coreComponents/constitutive/fluid/MultiFluidBase.hpp

```diff
@@ -339,6 +478,322 @@ class MultiFluidBase : public ConstitutiveBase
+    }
+  }
+}
+
+template< integer maxNumComp, integer maxNumPhase >
+GEOSX_HOST_DEVICE
+void
+MultiFluidBase::KernelWrapper::
+  convertToMassFractions( real64 const (&dCompMoleFrac_dCompMassFrac)[maxNumComp][maxNumComp],
+                          real64 const (&phaseMolecularWeight)[maxNumPhase],
+                          real64 const (&dPhaseMolecularWeight_dPres)[maxNumPhase],
+                          real64 const (&dPhaseMolecularWeight_dTemp)[maxNumPhase],
+                          real64 const (&dPhaseMolecularWeight_dGlobalCompFrac)[maxNumPhase][maxNumComp],
+                          PhaseProp::SliceType const phaseFrac,
+                          PhaseComp::SliceType const phaseCompFrac,
+                          arraySlice2d< real64, multifluid::USD_PHASE_DC - 2 > const dPhaseDens_dGlobalCompFrac,
+                          arraySlice2d< real64, multifluid::USD_PHASE_DC - 2 > const dPhaseVisc_dGlobalCompFrac ) const
+{
+  integer const numPhase = numPhases();
+  integer const numComp = numComponents();
+
+  // 1. Convert phase fractions (requires two passes)
+
+  real64 totalMass{};
+  real64 dTotalMass_dP{};
+  real64 dTotalMass_dT{};
+  real64 dTotalMass_dC[maxNumComp]{};
+
+  // 1.1. Compute mass of each phase and total mass (on a 1-mole basis)
+  for( integer ip = 0; ip < numPhase; ++ip )
+  {
+
+    bool const phaseExists = (phaseFrac.value[ip] > 0);
+    if( !phaseExists )
+    {
+      continue;
+    }
+
+    real64 const nu = phaseFrac.value[ip];
+
+    phaseFrac.value[ip] *= phaseMolecularWeight[ip];
+    phaseFrac.dPres[ip] = phaseFrac.dPres[ip] * phaseMolecularWeight[ip] + nu * dPhaseMolecularWeight_dPres[ip];
+    phaseFrac.dTemp[ip] = phaseFrac.dTemp[ip] * phaseMolecularWeight[ip] + nu * dPhaseMolecularWeight_dTemp[ip];
+
+    totalMass += phaseFrac.value[ip];
+    dTotalMass_dP += phaseFrac.dPres[ip];
+    dTotalMass_dT += phaseFrac.dTemp[ip];
+
+    for( integer jc = 0; jc < numComp; ++jc )
+    {
+      phaseFrac.dComp[ip][jc] =
+        phaseFrac.dComp[ip][jc] * phaseMolecularWeight[ip] + nu * dPhaseMolecularWeight_dGlobalCompFrac[ip][jc];
+      dTotalMass_dC[jc] += phaseFrac.dComp[ip][jc];
+    }
+  }
+
+  // 1.2. Normalize to get mass fractions
+  real64 const totalMassInv = 1.0 / totalMass;
+  for( integer ip = 0; ip < numPhase; ++ip )
+  {
+
+    bool const phaseExists = (phaseFrac.value[ip] > 0);
+    if( !phaseExists )
+    {
+      continue;
+    }
+
+    phaseFrac.value[ip] *= totalMassInv;
+    phaseFrac.dPres[ip] = ( phaseFrac.dPres[ip] - phaseFrac.value[ip] * dTotalMass_dP ) * totalMassInv;
+    phaseFrac.dTemp[ip] = ( phaseFrac.dTemp[ip] - phaseFrac.value[ip] * dTotalMass_dT ) * totalMassInv;
+
+    for( integer jc = 0; jc < numComp; ++jc )
+    {
+      phaseFrac.dComp[ip][jc] = ( phaseFrac.dComp[ip][jc] - phaseFrac.value[ip] * dTotalMass_dC[jc] ) * totalMassInv;
+    }
+  }
+
+  // 2. Convert phase compositions
```

Comment:
Same question, would it be better to make this into 3 smaller, more focused functions (at the cost of having to remember to call all 3 at in each model)?

Reply:
Yes, I implemented what you suggested, with three more focused functions doing just one specific task.